packages/plugins/typescript/apollo-client-helpers/src/index.ts

```diff
@@ -0,0 +1,84 @@
+import { Types, PluginValidateFn, PluginFunction } from '@graphql-codegen/plugin-helpers';
+import { GraphQLSchema, isInterfaceType, isObjectType } from 'graphql';
+import { extname } from 'path';
+import { ApolloClientHelpersConfig } from './config';
+
+export const plugin: PluginFunction<ApolloClientHelpersConfig> = (
+  schema: GraphQLSchema,
+  documents: Types.DocumentFile[],
+  config: ApolloClientHelpersConfig
+) => {
+  const results: Types.ComplexPluginOutput[] = [];
+
+  results.push(generateTypePoliciesSignature(schema, config));
+
+  return {
+    prepend: results.reduce((prev, r) => [...prev, ...(r.prepend || [])], []),
+    append: results.reduce((prev, r) => [...prev, ...(r.append || [])], []),
+    content: results.map(r => r.content).join('\n'),
+  };
+};
+
+function generateTypePoliciesSignature(
+  schema: GraphQLSchema,
+  config: ApolloClientHelpersConfig
+): Types.ComplexPluginOutput {
+  const typeMap = schema.getTypeMap();
+  const perTypePolicies: string[] = [];
+  const typedTypePolicies = Object.keys(typeMap).reduce((prev, typeName) => {
+    const type = typeMap[typeName];
+
+    if (!typeName.startsWith('__') && (isObjectType(type) || isInterfaceType(type))) {
+      const fieldsNames = Object.keys(type.getFields()).filter(f => !f.startsWith('__'));
+      const keySpecifierVarName = `${typeName}KeySpecifier`;
+      const fieldPolicyVarName = `${typeName}FieldPolicy`;
+
+      perTypePolicies.push(
+        `export type ${keySpecifierVarName} = (${fieldsNames
+          .map(f => `'${f}'`)
+          .join(' | ')} | ${keySpecifierVarName})[];`
+      );
+
+      perTypePolicies.push(`export type ${fieldPolicyVarName} = {
+${fieldsNames.map(fieldName => `\t${fieldName}?: FieldPolicy<any> | FieldReadFunction<any>`).join(',\n')}
+};`);
+
+      return {
+        ...prev,
+        [typeName]: `{
+\t\tkeyFields?: false | ${keySpecifierVarName} | (() => undefined | ${keySpecifierVarName}),
```

Comment:
and option for forcing `keyFields` to be defined would be sweet 👍

Reply:
except for possibly if one of the fields is named `id` as that's automatically picked up by apollo? not sure